src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs

```diff
@@ -99,6 +99,10 @@ public unsafe sealed override long Length
             }
         }
 
+        // in case of concurrent incomplete reads, there can be multiple threads trying to update the position
+        // at the same time. That is why we are using Interlocked here.
+        internal void OnIncompleteRead(int expectedBytesRead, int actualBytesRead) => Interlocked.Add(ref _filePosition, actualBytesRead - expectedBytesRead);
```

Comment:
@stephentoub I've used `Interlocked` here as the new `IncompleteReadCantSetPositionBeyondEndOfFile` test was simply failing without it. I know that we are never going to be 100% thread safe, but the alternative was to call `Length` which would result in an extra sys-call.

```cs
_filePosition = Math.Min(Length, _filePosition - expectedBytesRead + actualBytesRead);
```